### PR TITLE
fix: Fix registry proto

### PR DESCRIPTION
**Pull request by kevjumba**
**What this PR does / why we need it**:
Fixes the issue of registry proto caching not working at all. Added refreshes so calling any retrieval getter like list_entities actually looks into the cache to refresh before retrieving possible stale cache entities(quite a major bug that was left in the code). 

**Which issue(s) this PR fixes**:

Fixes #

### CODE REVIEW

File: go/internal/feast/registry.go [hunk 1]
```diff
@@ -28,7 +29,7 @@ type Registry struct {
 	cachedFeatureViews         map[string]map[string]*core.FeatureView
 	cachedOnDemandFeatureViews map[string]map[string]*core.OnDemandFeatureView
 	cachedRequestFeatureViews  map[string]map[string]*core.RequestFeatureView
-
+	cachedRegistry             *core.Registry
```
Comment (felixwang9817) [inline go/internal/feast/registry.go:32]: I thought the plan was to only keep the cached registry proto, just as in Python - what's the reason for keeping cachedEntities and other maps?

File: go/internal/feast/registry.go [hunk 2]
```diff
@@ -228,89 +248,113 @@ func (r *Registry) listOnDemandFeatureViews(project string) []*OnDemandFeatureVi
...
 
-func (r *Registry) getEntity(project, entityName string) (*Entity, error) {
+func (r *Registry) getEntity(project, entityName string, allowCache bool) (*Entity, error) {
+	_, err := r.getRegistryProto(allowCache)
+	if err != nil {
+		return nil, err
+	}
 	if entities, ok := r.cachedEntities[project]; !ok {
-		return nil, errors.New(fmt.Sprintf("project %s not found in getEntity", project))
+		return nil, fmt.Errorf("project %s not found in getEntity", project)
```
Comment (felixwang9817) [inline go/internal/feast/registry.go:275]: nit: if we keep this `cachedEntities` approach, this error message isn't super descriptive

maybe something like `no cached entities found for project %s`?

File: go/internal/feast/registry.go [hunk 3]
```diff
@@ -228,89 +248,113 @@ func (r *Registry) listOnDemandFeatureViews(project string) []*OnDemandFeatureVi
...
-			return nil, errors.New(fmt.Sprintf("onDemandFeatureView %s not found inside project %s", onDemandFeatureViewName, project))
+			return nil, fmt.Errorf("onDemandFeatureView %s not found inside project %s", onDemandFeatureViewName, project)
 		} else {
 			return NewOnDemandFeatureViewFromProto(onDemandFeatureViewProto), nil
 		}
 	}
 }
 
-func (r *Registry) getRequestFeatureView(project, requestFeatureViewName string) (*RequestFeatureView, error) {
+func (r *Registry) getRequestFeatureView(project, requestFeatureViewName string, allowCache bool) (*RequestFeatureView, error) {
```
Comment (achals) [inline go/internal/feast/registry.go:333]: I really don't like this `allowCache` bool that we pass around, even in the python implementation. For the go server can we always read the in-memory version `cachedRegistry`, and just support a different couple of methods - one to start a goroutine that refreshes on an interval, and another to force a refresh?

File: go/internal/feast/registry.go (hunk 1 above)
Comment (kevjumba) [reply go/internal/feast/registry.go:32]: It turns out we use the map in getEntity() and each of the other get Functionality. I think it makes sense now to generate the map before hand. 
Specifically, this makes it much faster to retrieve entities individually. Before we would compare every entity in the proto for either the project or the name but now we can just use the map instead.

File: go/internal/feast/registry.go [hunk 4]
```diff
@@ -171,17 +179,21 @@ func (r *Registry) listEntities(project string) []*Entity {
...
 
-func (r *Registry) listFeatureViews(project string) []*FeatureView {
-	if featureViewProtos, ok := r.cachedFeatureViews[project]; !ok {
-		return []*FeatureView{}
+func (r *Registry) listFeatureViews(project string, allowCache bool) ([]*FeatureView, error) {
+	_, err := r.getRegistryProto(allowCache)
+	if err != nil {
+		return nil, err
+	}
+	if cachedFeatureViews, ok := r.cachedFeatureViews[project]; !ok {
```
[the remark this answers is not in the document]
Comment (kevjumba) [reply go/internal/feast/registry.go:187]: Fixed.

File: go/internal/feast/registry.go (hunk 1 above)
Comment (kevjumba) [reply go/internal/feast/registry.go:32]: I think this design is probably better and removes latency from registry metadata retrieval and offloads the work of loading the featureviews to the actual cache refresh. 

Specifically, each time GetEntity() or listEntities() is called, it loops through every entity in the registry and compares the project name and entity name to retrieve the specified entities. I think it makes more sense to just keep it as is as a map so then retrieving them is just O(n) in the number of entities/project that matches the name.
What do you think @achals

File: go/internal/feast/registry.go (hunk 3 above)
Comment (kevjumba) [reply go/internal/feast/registry.go:333]: So what you're saying is just remove allowCache altogether?

Comment (achals) [reply go/internal/feast/registry.go:333]: I'd want to yeah, I don't think the interface needs it

File: go/internal/feast/registry.go (hunk 1 above)
Comment (achals) [reply go/internal/feast/registry.go:32]: I think exploding out into individual maps is fine in that case. It probably makes access easier (but we should make sure that updates to the individual fields are atomic).

Comment (kevjumba) [reply go/internal/feast/registry.go:33]: Yup, I locked the load functionality to make sure it's atomic.

File: go/internal/feast/registry.go (hunk 3 above)
Comment (kevjumba) [reply go/internal/feast/registry.go]: Ok I removed the allowcache interface and added a go routine. We already have a force refresh function and now we always read from the in memory cache.

File: go/internal/feast/featurestore.go [hunk 5]
```diff
@@ -636,26 +649,32 @@ func (fs *FeatureStore) dropUnneededColumns(onlineFeaturesResponse *serving.GetO
 	}
 }
 
-func (fs *FeatureStore) listFeatureViews(allowCache, hideDummyEntity bool) []*FeatureView {
-	featureViews := fs.registry.listFeatureViews(fs.config.Project)
+func (fs *FeatureStore) listFeatureViews(hideDummyEntity bool) ([]*FeatureView, error) {
```
Comment (achals) [inline go/internal/feast/featurestore.go:652]: How come we have a `FeatureStore.listFeatureViews` but no `FeatureStore.listRequestFeatureViews`?

File: go/internal/feast/registry.go [hunk 6]
```diff
@@ -28,9 +30,10 @@ type Registry struct {
 	cachedFeatureViews         map[string]map[string]*core.FeatureView
 	cachedOnDemandFeatureViews map[string]map[string]*core.OnDemandFeatureView
 	cachedRequestFeatureViews  map[string]map[string]*core.RequestFeatureView
-
+	cachedRegistry             *core.Registry
 	cachedRegistryProtoCreated time.Time
```
Comment (achals) [inline go/internal/feast/registry.go:34]: Maybe this should be `cachedRegistryProtoUpdated`? Since it's when the cached registry was last updated.

**Review by achals — APPROVED** [review]
small nits but looks fine

File: go/internal/feast/featurestore.go (hunk 5 above)
Comment (kevjumba) [reply go/internal/feast/featurestore.go:652]: Added.

File: go/internal/feast/registry.go (hunk 6 above)
Comment (kevjumba) [reply go/internal/feast/registry.go:34]: Done.

File: go/internal/feast/featurestore.go [hunk 7]
```diff
@@ -654,14 +654,17 @@ func (fs *FeatureStore) listFeatureViews(hideDummyEntity bool) ([]*FeatureView,
...
 	return featureViews, nil
 }
 
+func (fs *FeatureStore) listRequestFeatureViews() ([]*RequestFeatureView, error) {
+	requestFeatureViews, err := fs.registry.listRequestFeatureViews(fs.config.Project)
+	if err != nil {
+		return requestFeatureViews, err
+	}
+	return requestFeatureViews, nil
+}
```
Comment (achals) [inline go/internal/feast/featurestore.go:666]: can't this just be 
```
	return fs.registry.listRequestFeatureViews(fs.config.Project)
```

Comment (kevjumba) [reply go/internal/feast/featurestore.go:666]: Ah originally I added the hideDummyEntity but that's irrelevant here.

**Review by achals — APPROVED** [review]
/lgtm

**Pull request merged**